cloud/services/subnets/subnets.go

```diff
@@ -58,100 +47,100 @@ func (s *Service) getExisting(ctx context.Context, rgName string, spec *Spec) (*
 }
 
 // Reconcile gets/creates/updates a subnet.
-func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
-	subnetSpec, ok := spec.(*Spec)
-	if !ok {
-		return errors.New("Invalid Subnet Specification")
-	}
-	existingSubnet, err := s.getExisting(ctx, s.Scope.Vnet().ResourceGroup, subnetSpec)
-	if err == nil {
-		// subnet already exists, update the spec and skip creation
-		var subnet *infrav1.SubnetSpec
-		if subnetSpec.Role == infrav1.SubnetControlPlane {
-			subnet = s.Scope.ControlPlaneSubnet()
-		} else if subnetSpec.Role == infrav1.SubnetNode {
-			subnet = s.Scope.NodeSubnet()
-		} else {
+func (s *Service) Reconcile(ctx context.Context) error {
+	for i, subnetSpec := range s.Scope.SubnetSpecs() {
+		existingSubnet, err := s.getExisting(ctx, s.Scope.Vnet().ResourceGroup, subnetSpec)
+		if err == nil {
```

Comment:
What do you think about the following refactor to clean things up a bit:

```go
for i, subnetSpec := range s.Scope.SubnetSpecs() {
	existingSubnet, err := s.getExisting(ctx, s.Scope.Vnet().ResourceGroup, subnetSpec)
	switch {
		case err != nil && !azure.ResourceNotFound(err):
			// not found
 			return errors.Wrapf(err, "failed to get subnet %s", subnetSpec.Name)
		case err != nil:
			// stuff for existing subnet
		case !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()):
			// if the vnet is unmanaged, we expect all subnets to be created as well
			return fmt.Errorf("vnet was provided but subnet %s is missing", subnetSpec.Name
		default:
			// stuff create new subnet
	}
```

Reply:
Just wanted to call out: "why does is managed need `s.Scope.ClusterName()`?" Seems odd to have to pass in `s.Scope.ClusterName()` when calling a func off of hierarchical descendant of `s.Scope`.

Reply:
I think it's because IsManaged is on the infra VnetSpec, not the scope itself.
this could also read:

```
vnet := s.Scope.Vnet() // return type *infrav1.VnetSpec
if !vnet.IsManagedBy(s.Scope.ClusterName()) {
 ...
} 
```

Reply:
we could change this into s.Scope.IsVnetManaged() though

Agree it seems a bit odd like this

Reply:
made all the requested changes,  please take a look again :) thanks for all inputs!